snapshots/devmapper/pool_device.go

```diff
@@ -0,0 +1,323 @@
+}
+
+// CreateThinDevice creates new devmapper thin-device with given name and size.
+// Device ID for thin-device will be allocated from metadata store.
+// If allocation successful, device will be activated with /dev/mapper/<deviceName>
+func (p *PoolDevice) CreateThinDevice(ctx context.Context, deviceName string, virtualSizeBytes uint64) error {
+	info := &DeviceInfo{
+		Name:  deviceName,
+		Size:  virtualSizeBytes,
+		State: Unknown,
+	}
+
+	// Save initial device metadata and allocate new device ID from store
+	if err := p.metadata.AddDevice(ctx, info); err != nil {
+		return errors.Wrapf(err, "failed to save initial metadata for new thin device %q", deviceName)
+	}
+
+	// Create thin device
+	if err := p.transition(ctx, deviceName, Creating, Created, func() error {
+		return dmsetup.CreateDevice(p.poolName, info.DeviceID)
+	}); err != nil {
+		return errors.Wrapf(err, "failed to create new thin device %q (dev: %d)", info.Name, info.DeviceID)
+	}
+
+	// Activate thin device
+	if err := p.transition(ctx, deviceName, Activating, Activated, func() error {
+		return dmsetup.ActivateDevice(p.poolName, info.Name, info.DeviceID, info.Size, "")
+	}); err != nil {
+		return errors.Wrapf(err, "failed to activate new thin device %q (dev: %d)", info.Name, info.DeviceID)
+	}
+
+	return nil
+}
+
+// CreateSnapshotDevice creates and activates new thin-device from parent thin-device (makes snapshot)
+func (p *PoolDevice) CreateSnapshotDevice(ctx context.Context, deviceName string, snapshotName string, virtualSizeBytes uint64) error {
+	baseInfo, err := p.metadata.GetDevice(ctx, deviceName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to query device metadata for %q", deviceName)
+	}
+
+	isActivated := baseInfo.State == Activated
+
+	// Suspend thin device if it was activated previously
+	if isActivated {
+		if err := p.transition(ctx, baseInfo.Name, Suspending, Suspended, func() error {
+			return dmsetup.SuspendDevice(baseInfo.Name)
+		}); err != nil {
+			return errors.Wrapf(err, "failed to suspend device %q", baseInfo.Name)
+		}
+	}
+
+	snapInfo := &DeviceInfo{
+		Name:       snapshotName,
+		Size:       virtualSizeBytes,
+		ParentName: deviceName,
+		State:      Unknown,
+	}
+
+	// Save snapshot metadata and allocate new device ID
+	if err := p.metadata.AddDevice(ctx, snapInfo); err != nil {
+		return errors.Wrapf(err, "failed to save initial metadata for snapshot %q", snapshotName)
+	}
+
+	// Create thin device snapshot
+	if err := p.transition(ctx, snapInfo.Name, Creating, Created, func() error {
+		return dmsetup.CreateSnapshot(p.poolName, snapInfo.DeviceID, baseInfo.DeviceID)
+	}); err != nil {
+		return errors.Wrapf(err,
+			"failed to create snapshot %q (dev: %d) from %q (dev: %d, activated: %t)",
+			snapInfo.Name,
+			snapInfo.DeviceID,
+			baseInfo.Name,
+			baseInfo.DeviceID,
+			isActivated)
+	}
+
+	if isActivated {
+		// Resume base thin-device
+		if err := p.transition(ctx, baseInfo.Name, Resuming, Resumed, func() error {
```

Comment:
maybe we can put it into `defer` when fail to create thin device snapshot  

is it safe to suspend device when we make it committed? We cannot create child snapshot on active snapshot.

Reply:
> maybe we can put it into `defer` when fail to create thin device snapshot

You mean call resume even if `CreateSnapshot` failed?

> is it safe to suspend device when we make it committed, right?

yes

Reply:
> You mean call resume even if `CreateSnapshot` failed?

Yes, it can make sure that resume can works well.

So if we can suspend the device when we commit a snapshot, the `CreateSnapshotDevice` can just create snapshot device on the suspend device without `suspend and resume` action. does it make senses?

Reply:
> Yes, it can make sure that resume can works well.

Yep, I can do that.

> So if we can suspend the device when we commit a snapshot, the `CreateSnapshotDevice` can just create snapshot device on the suspend device without `suspend and resume` action. does it make senses?

Technically suspend/resume never called as of now, because `containerd` doesn't take snapshot of an active snapshot (but it [will in 1.3](https://github.com/containerd/containerd/issues/633)). But I still would like to keep suspend/resume code, `PoolDevice.CreateSnapshotDevice` should work correctly in both cases (suspend is not forced by dm, so it's legal to take snapshots of an active device, but this may cause corruptions (see [thin-provisioning.txt](https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt)), which we don't want) and this might be needed in close future.

Reply:
cool. thanks for reminding me about `clone` interface. sounds good to me.